### PR TITLE
AutoRegisteringObjectGraphType can register methods

### PR DESCRIPTION
**Pull request by dnsfour**
Hello. I did not understand why registration of methods was not implemented for AutoRegistraionObjectGraphType and decided to make my own implementation. If this is ok, then I propose to consider introducing a new class Optional<T> for the parameters of such methods, which will indicate whether values have been passed.

Example:
```
UpdatePerson(int id, Optional<string> name, Optional<int?> age){
   var person = getPerson(id);
   if(name != null) person.Name = name;
   if(age != null) person.Age = age;
   UpdatePerson(person);
}
```

### CODE REVIEW

**Comment by Shane32** [comment]
Just FYI, I've written a fairly complex version of this functionality within PR #1376 in a separate but complementary library.  For instance, it can handle a method signature like this:

```csharp
[Description("Updates the specified person's name and optionally their age")]
public static Task<Person> Update([FromSource] int personId, [FromServices] MyDbContext dbContext, [Required] string name, int? age)
{
    var user = await dbContext.Persons.Find(personId);
    user.Name = name;
    if (age.HasValue) user.Age = age;
    await dbContext.SaveChanges();
    return user;
}
```

You can add scoped services in the constructor, or within the method signature as seen above.  You can specify which methods can run asynchronously, and which ones must run synchronously.

I would like to infer if the argument is optional via nullable reference types, but have not worked on that yet.

As you'll see, I have not polished it, but I do use it internally and hope to sometime add tests & documentation, and publish it as a nuget package.

There's also the GraphQL conventions project which also contains similar functionality - see https://github.com/graphql-dotnet/conventions - but I have not used it.

**Comment by dnsfour** [comment]
Let me explain my motivation:
1) My colleagues are complaining about a lot of code that goes with graphql.
2) With optional parameters, there is no need to get the whole entity to update one field, and I think this is important for point 1)
3) Source objects become adapters to entities, which must be wrapped each time in the resolve method manually, while services and data can be manually or ActivatorUtils.CreateInstance() injected. So you can do without additional attributes.


```cs
    public class PersonData
    {
        public int Id { get; set; }
        public int Name { get; set; }
    }

    /// <summary>Query root object</summary>
    public class Query
    {
        private readonly PersonStore personStore;
        private readonly IServiceProvider sp;

        public Query(PersonStore personStore, IServiceProvider sp)
        {
            this.personStore = personStore;
            this.sp = sp;
        }

        public async Task<Person> Person(int id)
        {
            var personData = await personStore.Read(id);
            return ActivatorUtilities.CreateInstance<Person>(sp, personData);
        }
    }

    /// <summary>Source for ObjectGraphType<Person></summary>
    public class Person
    {
        private readonly PersonData person;
        private readonly IDataLoaderContext dataLoader;
        private readonly PersonStore personStore;

        public Person(PersonData data, IDataLoaderContext dataLoader, PersonStore personStore)
        {
            this.person = data;
            this.dataLoader = dataLoader;
            this.personStore = personStore;
        }

        public int Id { get; set; }
        public string Name { get; set; }

        public Task<Person> Relativies(int skip, int take) => dataLoader.LoadBatch(...).LoadKey(data.Id);
        public Task<Person> Parents(IResolveFieldContext ctx) => dataLoader.LoadBatch(...).LoadKey(data.Id);
    }
```

**Comment by Shane32** [comment]
Interesting; I had not thought of that.  PR #1376 does not work this way, although it probably could be adapted to do so quite easily.  I not sure I follow how this design applies to changes to the database.  In other words, typically changes only occur in mutations, so wouldn't a separate object be required for mutations vs queries?  Perhaps you could share some more samples of how you anticipate this wires up.

**Comment by dnsfour** [comment]
Right now i guess only to do parsing start of query
Example with another way to pass services and data into source:
```cs
	                //arrange
			var de = new DocumentExecuter();
			var spMock = new Mock<IServiceProvider>();
			var sp = spMock.Object;
			var schema = new Schema {
				Query = new AutoRegisteringObjectGraphType<Query>(true),
				Mutation = new AutoRegisteringObjectGraphType<Mutation>(true)
			};
			var query = "mutation { ... }";
			//act
			var result = await de.ExecuteAsync(_ => {
				_.Schema = schema;
				_.Query = query;
				_.Root = query.TrimStart().StartsWith("mutation") ? sp.GetRequiredService(typeof(Mutation))
					: query.TrimStart().StartsWith("query") ? sp.GetRequiredService(typeof(Query))
					: null;
			});
		}


		public class ServicesContext {
			private readonly IServiceProvider sp;
			public ServicesContext(IServiceProvider sp) => this.sp = sp;

			private PersonStore personStore;
			public PersonStore PersonStore => personStore ??= sp.GetRequiredService<PersonStore>();
                        //...
		}

		// source of ObjectGraphType<Mutation>
		public class Mutation {
			private readonly ServicesContext services;

			public Mutation(ServicesContext services) {
				this.services = services;
			}

			public async Task<Person> UpdatePerson(int id, Optional<string> name) {
				var personData = await services.PersonStore.Read(id);
				if (name != null) personData.Name = name;
                                await services.PersonStore.Write(personData);
				return new Person(services, personData);
			}
		}

		// source of ObjectGraphType<Person>
		public class Person {
			private readonly ServicesContext services;
			private readonly PersonData data;

			public Person(ServicesContext services, PersonData data) {
				this.services = services;
				this.data = data;
			}

			//...
		}
```

**Comment by dnsfour** [comment]
And I also think that the new c#9 Record type should help in passing services and data:
```cs
return person with { Data = personData }
...
public record Person {
 public PersonData Data { get; init; }
 constructor(PersonStore store) {...}
}
```

**Comment by joemcbride** [comment]
Just noting here that schema-first syntax allows you to write your GraphQL types exactly as described.  It uses the [AccessorFieldResolver](https://github.com/graphql-dotnet/graphql-dotnet/blob/a35b22cc0c9ca4c8387af5daaf71635b56008ee4/src/GraphQL/Reflection/AccessorFieldResolver.cs#L6-L33) internally to accomplish this.

https://graphql-dotnet.github.io/docs/getting-started/introduction#schema-first-nested-types

https://github.com/graphql-dotnet/graphql-dotnet/blob/a35b22cc0c9ca4c8387af5daaf71635b56008ee4/src/GraphQL/Utilities/TypeConfig.cs#L43-L55

https://github.com/graphql-dotnet/graphql-dotnet/blob/a35b22cc0c9ca4c8387af5daaf71635b56008ee4/src/GraphQL/Reflection/AccessorFieldResolver.cs#L6-L33

**Comment by joemcbride** [comment]
Slightly updated for how schema-first would handle it:

```gql
type Person {
    Id: ID!
    name: String!
    relativies(skip: Int!, take: Int!): [Person!]!
    parents: [Person!]!
}

type Query {
    person(id: ID!): Person
}
```

```csharp
public class PersonData
{
    public int Id { get; set; }
    public int Name { get; set; }
}

/// <summary>Query root object</summary>
public class Query
{
    private readonly PersonStore personStore;

    public Query(PersonStore personStore)
    {
        this.personStore = personStore;
    }

    public async Task<Person> Person(int id)
    {
        return await personStore.Read(id);
    }
}

/// <summary>Source for ObjectGraphType<Person></summary>
public class Person
{
    private readonly IDataLoaderContext dataLoader;
    private readonly PersonStore personStore;

    public Person(IDataLoaderContext dataLoader, PersonStore personStore)
    {
        this.data = data;
        this.dataLoader = dataLoader;
        this.personStore = personStore;
    }

    // these are not required
    // public int Id(PersonData data) => data.Id;
    // public string Name(PersonData data) => data.Name;

    public Task<Person[]> Relativies(int skip, int take) => dataLoader.LoadBatch(...).LoadKey(data.Id);
    public Task<Person[]> Parents(IResolveFieldContext ctx) => dataLoader.LoadBatch(...).LoadKey(data.Id);
}
```

```csharp
var schema = Schema.For("..." _ =>
{
    _.Types.Include<Query>();
    _.Types.Include<Person>();
});

var json = await schema.ExecuteAsync(_ =>
{
    _.Query = "{ person(id: ""123"") { id name parents { name } } }";
    _. RequestServices = ...
});

Console.WriteLine(json);
```

File: src/GraphQL.ApiTests/ApiApprovalTests.PublicApi.GraphQL.approved.txt.orig [hunk 1]
```diff
@@ -0,0 +1,3130 @@
+namespace GraphQL
```
Comment (sungam3r) [inline src/GraphQL.ApiTests/ApiApprovalTests.PublicApi.GraphQL.approved.txt.orig:1]: Please remove this file.

File: src/GraphQL.Tests/Types/ComplexGraphTypeTests.cs [hunk 2]
```diff
@@ -82,11 +90,16 @@ public void auto_register_object_graph_type()
         {
             GraphTypeTypeRegistry.Register<Money, AutoRegisteringObjectGraphType<Money>>();
 
-            var type = new AutoRegisteringObjectGraphType<TestObject>(o => o.valuePair, o => o.someEnumerable);
+            var type = new AutoRegisteringObjectGraphType<TestObject>(
+                true,
+                o => o.valuePair,
+                o => o.someEnumerable,
+                o => o.someIntMethod2(default, default, default)
```
Comment (sungam3r) [inline src/GraphQL.Tests/Types/ComplexGraphTypeTests.cs:97]: it looks a little strange

File: src/GraphQL/Resolvers/MethodResolver.cs [hunk 3]
```diff
@@ -0,0 +1,85 @@
...
+
+namespace GraphQL.Resolvers
+{
+    /// <summary>
+    /// Experimental. Probably should be a part of <see cref="NameFieldResolver"/>
+    /// It has <see cref="CreateQueryArguments"/> for field registration
+    /// </summary>
+    internal class MethodResolver : IFieldResolver
+    {
+        public static readonly ConcurrentDictionary<MethodInfo, Func<object, object[], object>> cache = new ConcurrentDictionary<MethodInfo, Func<object, object[], object>>();
```
Comment (sungam3r) [inline src/GraphQL/Resolvers/MethodResolver.cs:16]: public?

File: src/GraphQL/TypeExtensions.cs [hunk 4]
```diff
@@ -94,6 +103,8 @@ public static string GraphQLName(this Type type)
                 : typeName;
         }
 
+
+
```
Comment (sungam3r) [inline src/GraphQL/TypeExtensions.cs:107]:
```suggestion
```

File: src/GraphQL/Types/Composite/AutoRegisteringObjectGraphType.cs [hunk 5]
```diff
@@ -18,18 +20,34 @@ namespace GraphQL.Types
...
         /// </summary>
-        public AutoRegisteringObjectGraphType() : this(null) { }
+        public AutoRegisteringObjectGraphType() : this(false, null) { }
 
         /// <summary>
         /// Creates a GraphQL type from <typeparamref name="TSourceType"/> by specifying fields to exclude from registration.
         /// </summary>
-        /// <param name="excludedProperties"> Expressions for excluding fields, for example 'o => o.Age'. </param>
-        public AutoRegisteringObjectGraphType(params Expression<Func<TSourceType, object>>[] excludedProperties)
+        /// <param name="excludesMembers"> Expressions for excluding fields, for example 'o => o.Age' or 'o => o.Age(defaul, ...). </param>
```
Comment (sungam3r) [inline src/GraphQL/Types/Composite/AutoRegisteringObjectGraphType.cs:35]:
```suggestion
        /// <param name="excludesMembers"> Expressions for excluding fields, for example 'o => o.Age' or 'o => o.Age(default, ...). </param>
```

File: src/GraphQL/Types/Composite/AutoRegisteringObjectGraphType.cs [hunk 6]
```diff
@@ -18,18 +20,34 @@ namespace GraphQL.Types
...
+        public AutoRegisteringObjectGraphType(params Expression<Func<TSourceType, object>>[] excludesMembers): this(false, excludesMembers)
         {
-            AutoRegisteringHelper.SetFields(this, GetRegisteredProperties(), excludedProperties);
+        }
+
+        /// <summary>
+        /// Creates a GraphQL type from <typeparamref name="TSourceType"/> by specifying fields to exclude from registration.
+        /// </summary>
+        /// <param name="registerMethods"> Register methods of <see cref="GetRegisteredMethods"/> </param>
+        /// <param name="excludesMembers"> Expressions for excluding fields, for example 'o => o.Age' or 'o => o.Age(defaul, ...). </param>
```
Comment (sungam3r) [inline src/GraphQL/Types/Composite/AutoRegisteringObjectGraphType.cs:44]:
```suggestion
        /// <param name="excludesMembers"> Expressions for excluding fields, for example 'o => o.Age' or 'o => o.Age(default, ...). </param>
```

File: src/GraphQL/Types/Composite/AutoRegisteringObjectGraphType.cs [hunk 7]
```diff
@@ -38,6 +56,13 @@ protected virtual IEnumerable<PropertyInfo> GetRegisteredProperties()
                 .GetProperties(BindingFlags.Public | BindingFlags.Instance)
                 .Where(p => AutoRegisteringHelper.IsEnabledForRegister(p.PropertyType, true));
         }
+
+        protected virtual IEnumerable<MethodInfo> GetRegisteredMethods()
+        {
+            return typeof(TSourceType)
+                .GetMethods(BindingFlags.Public | BindingFlags.Instance | BindingFlags.DeclaredOnly)
```
Comment (sungam3r) [inline src/GraphQL/Types/Composite/AutoRegisteringObjectGraphType.cs:63]: Why DeclaredOnly?

File: src/GraphQL/Types/Composite/AutoRegisteringObjectGraphType.cs [hunk 8]
```diff
@@ -75,78 +101,102 @@ internal static class AutoRegisteringHelper
...
         }
 
-        internal static bool IsEnabledForRegister(Type propertyType, bool firstCall)
+        internal static bool IsEnabledForRegister(Type returnType, bool firstCall)
         {
-            if (propertyType == typeof(string)) return true;
+            if (returnType == typeof(string))
+                return true;
 
-            if (propertyType.IsValueType) return true; // TODO: requires discussion: Nullable<T>, enums, any struct
```
Comment (sungam3r) [inline src/GraphQL/Types/Composite/AutoRegisteringObjectGraphType.cs:118]: Please revert all code formatting changes. These changes are not relevant and only complicate the review.

**Review by sungam3r — CHANGES_REQUESTED** [review]
> I did not understand why registration of methods was not implemented for AutoRegistraionObjectGraphType

Most likely because method registration is not supported for regular GraphTypes. If you are adding this function for AutoRegisteringObjectGraphType, then the first thing to ask yourself is whether it is worth adding it to the regular GraphType APIs.

File: src/GraphQL.Tests/Types/ComplexGraphTypeTests.cs (hunk 2 above)
Comment (dnsfour) [reply src/GraphQL.Tests/Types/ComplexGraphTypeTests.cs:97]: Agree, but powerful to exclude overloaded methods

File: src/GraphQL/Resolvers/MethodResolver.cs (hunk 3 above)
Comment (dnsfour) [reply src/GraphQL/Resolvers/MethodResolver.cs:16]: Okay. I was not sure about the name and whether it should be separated from the NameFieldResolver.

File: src/GraphQL/Types/Composite/AutoRegisteringObjectGraphType.cs (hunk 7 above)
Comment (dnsfour) [reply src/GraphQL/Types/Composite/AutoRegisteringObjectGraphType.cs:63]: The graphql spec does not support inheritance and so there will be better schema match

File: src/GraphQL/Types/Composite/AutoRegisteringObjectGraphType.cs (hunk 8 above)
Comment (dnsfour) [reply src/GraphQL/Types/Composite/AutoRegisteringObjectGraphType.cs:118]: done

**Comment by dnsfour** [comment]
> first thing to ask yourself is whether it is worth adding it to the regular GraphType APIs.

Maybe I'm wrong, but I thought that AutoRegisteringObjectGraphType is not regular GraphType

**Comment by dnsfour** [comment]
> Just noting here that schema-first syntax allows you to write your GraphQL types exactly as described. It uses the AccessorFieldResolver internally to accomplish this.

I really missed it.
1) Reflection is used to call methods. This is slower than my implementation, I can improve
2) It seems that the schema-first approach increases the level of entry into graphql-dotnet for line developers, since you need to know the graphql syntax well.
3)  Schema-first means more code.

File: src/GraphQL/Resolvers/MethodResolver.cs (hunk 3 above)
Comment (dnsfour) [reply src/GraphQL/Resolvers/MethodResolver.cs]: > Okay. I was not sure about the name and whether it should be separated from the NameFieldResolver.

Or is it about cache? Fixed public #cache.

**Comment by dnsfour** [comment]
Sorry for wasting time.
I realized this was a bad idea and it would be better if I do a couple of PRs:

- Improve NameFieldResolver to optinal pass source, ResolveFieldContext and arguments to methodsS
- Add new Optional<T> class so that you can distinguish between updatePerson (id: 1 name: null) and updatePerson (id: 1 dateOfBirth: null)
- Do breaking change for AutoRegisteringObjectGraphType: rename GetRegisteredProperties to GetRegisterMembers, add to constructor arguments registerProperties registerMethods and depricate existing constructors.

If I still fit into the concept of regular graphtypes.

File: src/GraphQL/Types/Composite/AutoRegisteringObjectGraphType.cs (hunk 7 above)
Comment (sungam3r) [reply src/GraphQL/Types/Composite/AutoRegisteringObjectGraphType.cs:63]: It is all about implementation details here, not about spec.

**Comment by sungam3r** [comment]
> Maybe I'm wrong, but I thought that AutoRegisteringObjectGraphType is not regular GraphType

Yes but I talked about regular graph types and graphtype first syntax.

**Pull request closed without merging**